### PR TITLE
Add const generics to the HIR

### PR DESCRIPTION
**Pull request by varkor**
Split out from https://github.com/rust-lang/rust/pull/53645. This work is a collaborative effort with @yodaldevoid.

r? @eddyb

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -389,24 +389,33 @@ impl PathSegment {
     }
 }
 
+#[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
+pub struct ConstArg {
+    pub value: AnonConst,
+    pub span: Span,
```
Comment (petrochenkov) [inline src/librustc/hir/mod.rs]: This span is identical to the span in body of `AnonConst`, is it viable to use it instead?
Other `AnonConst`s don't create extra spans during lowering as far as can I see.

Comment (petrochenkov) [reply src/librustc/hir/mod.rs]: (The whole `ConstArg` could be refactored away in that case.)

Comment (petrochenkov) [reply src/librustc/hir/mod.rs]: Ugh, I see `GenericArg::span` wants to work in isolation without any contexts.

Comment (yodaldevoid) [reply src/librustc/hir/mod.rs]: Yeah, we went through the same thought process when this was implemented. If you can see a way to refactor this, I'd be all ears.

File: src/librustc_typeck/collect.rs [hunk 2]
```diff
@@ -1041,6 +1045,21 @@ fn generics_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> &'tcx ty
...
+                            param.span,
+                            "`Self` should not be the name of a regular parameter",
+                        );
+                    }
+
+                    tcx.sess.struct_span_err(
+                        param.span,
+                        "const generics in any position are currently unsupported",
+                    ).emit();
+                    tcx.sess.abort_if_errors();
```
Comment (petrochenkov) [inline src/librustc_typeck/collect.rs]: No need to abort, we can skip the parameter and continue.
This way we could e.g. test other things added in this PR like the uppercase lint.

**Comment by petrochenkov** [comment]
LGTM, r=me modulo removing abort / adding lint test (https://github.com/rust-lang/rust/pull/58503#discussion_r257449004).

I'll keep @eddyb assigned in case he wants to look.

**Comment by varkor** [comment]
@bors r=petrochenkov p=1

(Increasing priority because this unblocks other work.)

File: src/librustc_metadata/encoder.rs [hunk 3]
```diff
@@ -1684,6 +1707,11 @@ impl<'a, 'b, 'tcx> IndexBuilder<'a, 'b, 'tcx> {
                     let encode_info = IsolatedEncoder::encode_info_for_ty_param;
                     self.record(def_id, encode_info, (def_id, has_default));
                 }
+                hir::GenericParamKind::Const { .. } => {
+                    let def_id = self.tcx.hir().local_def_id(param.id);
+                    let encode_info = IsolatedEncoder::encode_info_for_const_param;
```
Comment (eddyb) [inline src/librustc_metadata/encoder.rs]: Maybe open an issue about unifying the param encoding in metadata, the same way `GenericParamDef` has common fields?

Comment (varkor) [reply src/librustc_metadata/encoder.rs]: Opened https://github.com/rust-lang/rust/pull/58581.

**Pull request merged**